lib/db_connection.ex

```diff
@@ -1208,275 +1236,192 @@ defmodule DBConnection do
+        fail(conn)
+        {nil, {kind, reason, stack, nil}}
     else
       result ->
-        result = {:ok, result}
-        commit(conn, log, opts, result)
+        {nil, conclude(conn, result)}
+    end
+  end
+  defp run_transaction(conn, {:failed, _}, fun, _meter, _opts) do
+    %DBConnection{conn_ref: conn_ref} = conn
+    try do
+      fun.(conn)
     catch
       :throw, {:rollback, ^conn_ref, reason} ->
-        result = {:error, reason}
-        rollback(conn, log, opts, result)
+        {nil, {:error, reason, nil}}
       kind, reason ->
-        result = {kind, reason, System.stacktrace()}
-        rollback(conn, log, opts, result)
+        stack = System.stacktrace()
+        {nil, {kind, reason, stack, nil}}
+    else
+      _result ->
+        {nil, {:error, :rollback, nil}}
+    end
+  end
+  defp run_transaction(conn, conn_info, fun, meter, opts) do
+    case run_begin(conn, conn_info, meter, opts) do
+      {:ok, res, meter} ->
+        %DBConnection{conn_ref: conn_ref} = conn
+        try do
+          log(meter, :transaction, :begin, nil, {:ok, res})
+          fun.(conn)
+        catch
+          :throw, {:rollback, ^conn_ref, reason} ->
+            run_rollback(conn, get_info(conn), reason, meter(opts), opts)
+          kind, reason ->
+            stack = System.stacktrace()
+            raised_rollback(conn, opts, kind, reason, stack)
+        else
+          result ->
+            run_commit(conn, get_info(conn), result, meter(opts), opts)
+        end
+      {:error, err, meter} ->
+        {:begin, {:error, err, meter}}
+      {kind, reason, stack, meter} ->
+        {:begin, {kind, reason, stack, meter}}
     end
   end
 
-  defp commit(conn, log, opts, result) do
+  defp fail(conn) do
     case get_info(conn) do
       {:transaction, conn_state} ->
-        conclude_meter(conn, conn_state, log, :handle_commit, opts, result)
-      {:failed, conn_state} ->
-        result = {:error, :rollback}
-        conclude_meter(conn, conn_state, log, :handle_rollback, opts, result)
-      :closed ->
-        {{:error, :rollback}, nil}
+        put_info(conn, :failed, conn_state)
+      _ ->
+        :ok
     end
   end
 
-  defp rollback(conn, log, opts, result) do
+  defp conclude(conn, result) do
     case get_info(conn) do
-      {trans, conn_state} when trans in [:transaction, :failed] ->
-        conclude_meter(conn, conn_state, log, :handle_rollback, opts, result)
-      :closed ->
-        {result, nil}
+      {:transaction, _} ->
+        {:ok, result, nil}
+      _ ->
+        {:error, :rollback, nil}
     end
   end
 
-  defp conclude_meter(conn, conn_state, nil, callback, opts, result) do
-    case handle(conn, conn_state, callback, opts, :idle) do
-      {:ok, _} ->
-        {result, nil}
-      error ->
-        {error, nil}
-    end
+  defp run_begin(conn, conn_info, meter, opts) do
+    meter = event(meter, :begin)
+    handle_trans(conn, conn_info, :handle_begin, meter, opts, :transaction)
```

Comment:
Transaction calls aren't wrapped in `run` calls with funs but could be as parameters align.